src/ir/datalog/program.h

```diff
@@ -101,12 +116,22 @@ class ArgumentType {
 
   std::string ToString() const { return absl::StrCat(kind_, name_); }
 
+  // This is needed because Argument has private members
+  // of type ArgumentType, RelationDeclaration has private
+  // members of type Argument, and RelationDeclaration
+  // appears in an absl::flat_hash_map in the implementation of
+  // DeclarationEnvironment.
+  template <typename H>
+  friend H AbslHashValue(H h, const ArgumentType& typ) {
+    return H::combine(std::move(h), typ.name_, typ.kind_);
+  }
+  // Equality is needed to use a RelationDeclaration in a flat_hash_set
   bool operator==(const ArgumentType& otherType) const {
-    // The name field is only compared for arguments of kind kCustom
-    // because the name field is only used for kCustom types (and
-    // not for kPrincipal or kNumber).
-    return this->kind_ == otherType.kind_ &&
-      this->kind_ == Kind::kCustom ? this->name_ == otherType.name_ : true;
+    return this->kind_ == otherType.kind_ && this->name_ == otherType.name_;
+  }
+  // This one is just for convenience
```

Comment:
The comment doesn't seem right. This is still useful?

Reply:
There is a comment for `==` to explain why it is needed, but `!=` isn't *strictly* needed for anything. It is just for convenience given that there is already a `==` operator. Since this comment is more confusing that helpful, I'll delete it.